## Probability 

Bu bölümde öğrenilenler şu şekilde olacaktır:

Probability Nedir?

Permutation (Permutasyon)

Combination (Kombinasyon)

Intersection, Unions and Complements (Kesişim, Birleşim ve Tamamlayıcı)

Independent and Dependent Events (Bağımsız ve Bağımlı Olaylar)

Conditional Probability (Şartlı olasılık)

Bayes Theorem (Bayes teoremi)

### Probability Nedir?

Olasılık bir şeyin olmasının veya olmamasının matematiksel değeri veya olabilirlik yüzdesi, değeridir.

Probability 0 ve 1 arasında değerler alabilir. 1 değeri olayın kesin olduğunu ve 0 değeri ise olayın gerçekleşmediğini ifade eder.

1 olayın kesin olduğunu 0 ise olayın gerçekleşmediğini gösterir.

Bir paranın yazı gelme olasılığı 0.5 dir. P(para_yazı) = 0.5 şeklinde gösterilir.

Parayı yazı mı turamı diyerek havaya atmaya trial(deneme) denir. 

Bu trial olayları birbirinde independent olaylardır. Yani birbirinden bağımsıdır. Mesela 5 kez zar atıyoruz. Atılan ilk 4 zarın hiçbir şekilde atılan son zara bir etkisi olmayacaktır. Dolayısıyla trial'lar birbirinden bağımsız olaylardır.

Experiment: her bir trial olayı experiment olarak adlandırılır.

Simple event: Ortaya çıkabilecek sonuçlar yani yazı yada tura simple event olarak adlandırılır.

Sample Space: tüm olabilecek simple event'lerin toplamıdır.

Örneğin:

    Mesela bir zar atalım.
    
    Zarı atmak experiment
    
    Simple events zarı atmanın sonucunda ortaya çıkabilecek sonuçlar 1,2,3,4,5,6 değerleri.
    
    Sample space => S = {1,2,3,4,5,6} 6 possible outcomes

### Permutation (Permutasyon)

Matematikte permütasyon, her sembolün sadece bir veya birkaç kez kullanıldığı sıralı bir dizidir.

Mesela 3 tane öğrenci olsun: ali, veli ve ayşe

Bu üç öğrenciyi kaç farklı şekilde yan yana sıralarız? 3! = 3.2.1 = 6

    ali, veli, ayşe

    ali, ayşe, veli

    veli, ayşe, ali

    veli, ali, ayşe

    ayşe, veli, ali

    ayşe, ali, veli

n elemanlık bir kümenin permutasyonu n!

n elemanlık bir kümenin r alana permutasyonu  (tekrarlamak yasak) nedir?

Şu formül ile hesaplanabilir:

<br>
<img src="p1.png" align = "left" style="width:200px;height:100px"/>

Örnek:

    Bir web sitesine üye olacaksınız ve 5 karakterli bir şifre belirleyeceksiniz.

    Karakterler küçük harf olmak zorunda yada 0-9 arası sayılar olabilir.

    Kullanılan bir karakter yada sayıyı tekrar kullanmak yoktur.

    Bu durumda kaç farklı şifre yaratabilirsiniz?

        n = 29(harf) + 10(sayılar) = 39

        r = 5 (5 tane boş alan olduğu için r=5'tir.)

        P = 39! / (39-5)! = 39! / (34!)
        
Bu işlemleri gerçekleştiren kodu yazalım.

In [1]:
import math
pay = math.factorial(39)
payda = math.factorial(34)
P = int(pay / payda)
print(P)

69090840


5 haneli bir şifre oluşturmamız 69090840 farklı şekilde olabilir. Bir insanın şifresini bulmamız için 69090840 kere deneme yapmamız gerekecektir.

<br>
<img src="p2.png" align = "left" style="width:200px;height:100px"/>

Yukarıda ki görseli açıklayayım. Toplamda 39 karakter bulunmaktadır. Bir karakter sadece 1 defa kullanılabilir. Dolayısıyla ilk karakterde 39 seçenek vardır. Sonrasında ilk karakterde kullanılan karakter dışlanır ve geriye 38 karakter kalır. 5 adet karakter yazılacağından 39.38.37.36.35 = 69090840 elde edilir. Yani aslında P(n,r) = n! / (n-r)! formülü bu işlemi yapmaktadır. 

Eger kullanılan bir karakter tekrar kullanılmaya izin verilseydi:

<br>
<img src="p3.png" align = "left" style="width:100px;height:70px"/><br><br><br><br><br>

formülü gerçekleştirilirdi. Yani:

<br>
<img src="p4.png" align = "left" style="width:200px;height:100px"/>

Kullanılan bir karakter istenilen adette kullanılabileceğinden her defasında 39 karakter seçeneği olacaktır. Dolayısıyla 39*39*39*39*39 olur ve buda 39^5 e eşittir.

In [2]:
# eger kullanılan bir karakter tekrar kullanılmaya izin verilseydi P = n^r
39**5

90224199

Bir karakterin istenen adette kullanılabilmesi durumunda 90224199 adet farklı şifre oluşturulabilir.

3 farklı pantolon, 4 farklı ceket ve 2 farklı gömleği olan kişi bir ceket bir gömlek ve bir pantolonu kaç farklı şekilde giyebilir?

    P = 3.4.2 = 24 

    yada formule göre: 3! / (3-1)! * 4! / (4-1)! * 2! / (2-1)! = 24

### Combination (Kombinasyon)

Kombinasyon, bir nesne grubu içerisinden sıra gözetmeksizin yapılan seçimlerdir.

Mesela 5 kişilik bir takımdan 2 kişi seçmek istiyoruz. Sıra gözetmeksizin 5 kişinden 2 kişi seçeriz.

<br>
<img src="p5.png" align = "left" style="width:200px;height:100px"/>

n = 5 ve r = 2 sonuç: 10 farklı şekilde seçeriz.

### Permutation vs Combination

Üç öğrenci: ali, veli ve ayşe. Sıralamak demek permutation'dur. İçinden öğrenci seçmek combination demektir.

Mesela 3 öğrenciyi 2 yere sıralayalım. 6 farklı şekilde yaparız.

    ali, veli

    ali, ayşe

    veli, ayşe

    veli, ali

    ayşe, veli

    ayşe, ali
    
Bu üç öğrenciden 2 öğrenci seçelim. Sonuç 3 çıkar Gördüğünüz gibi sıralamada ali, ayşe var aynı zamanda ayşe, ali var. Bu durum sıralama için farklı olsada seçme yani combination için aynı durumlardır.

p(3,2) = 3! / 1! = 6

c(3,2) = 3! / (1!*2!) = 3

### Intersection, Unions and Complements (Kesişim, Birleşim ve Tamamlayıcı)

Intersection(Kesişim): iki olayın aynı anda olmasıdır. A ve B

Unions(Birleşim): iki olayın ikisinden birinin olmasıdır. A or B

<br>
<img src="p6.png" align = "left" style="width:500px;height:600px"/>

disjoint: Kesişim olmadığını ifade eder.

### Independent and Dependent Events (Bağımsız ve Bağımlı Olaylar)

Independent(Bağımsız) Events:

İki olay birbirinden bağımsızdır. Yani bir event'in sonucu öbürünü etkilememektedir.

Mesela 5 kez zar atılacaktır. İlk 4 zar atma olayı gerçekleştirilmiştir. 5. zar atılınca sonucun 3 gelme olasılığı nedir? 
Cevap 1/6. 

Daha önceden atılan 4 zar 5. atılan zar ile hiç alakalı değildir. Yani bu olaylar birbiri ile independent'tir.

Dependent(Bağımlı) Events:

İki olay birbiri ile alakalıdır.

Mesela bir çanta olsun içerisinde 2 tane kırmızı 2 tane mavi top var.

Çanta içinden ard arda 2 kez top çekeceğim. 2. topun kırmızı olma olasılığı nedir?

Bu durumda çantadan top çekme olayları dependent'tir. Çünkü eğer ilk seferde kırmızı çekersem çantada 2 mavi 1 kırmızı kalır. Eğer ilk seferde mavi çekersem çantada 1 mavi 2 kırmızı kalır.

İndependent ve dependent durumları önemlidir çünkü bu durumları feature'ler arasında değerlendireceğiz.

### Conditional Probability (Şartlı olasılık)

Conditional Probability: Probability of event A given B. Bu ifade şu şekilde gösterilmektedir:

P(A|B)

B event'inin olma olasılığı A'ya bağlıdır. 

Örnek: Bir zar atıldığında üste gelen sayının 2 den büyük olduğu bilindiğine göre, zarın üst yüzüne gelen sayının çift sayı olma olasılığı kaçtır?

E = {1,2,3,4,5,6}

2 den büyük olma şartı (condition) B = {3,4,5,6}

B koşulu sağlandıktan sonra B kümesi içerisindeki çift olan sayılar {4,6}

Sonuç: (2/6) / (4/6) =  1/2

Burada P(A and B) = A ve B nin kesişimi {4,6} şeklindedir. Toplamda da 6 sayı var. 2/6'dır. 

P(B): B={3,4,5,6} dır ve 4 tane sayı vardır. Toplamda 6 sayı vardır. Dolayısıyla P(B)=4/6 dır.

<br>
<img src="p7.png" align = "left" style="width:300px;height:100px"/>

E = {1,2,3,4,5,6}
B = {3,4,5,6}
A = {2,4,6}

Örnek:

52 lik bir oyun kartı destesini karıştırıyoruz. Karıştırdıktan sonra kapalı bir şekilde masaya koyuyoruz. Sonra en üstten iki tane kart çekiyoruz ikisininde sinek gelme olasılığı nedir?

52/4 = 13 tane sinek kartı vardır.

P(sinek2| sinek1) = P(sinek1 ∩ sinek2) / P(sinek1)

P(sinek1) * P(sinek2| sinek1) = P(sinek1 ∩ sinek2) = 13/52 · 12/51 = 0.0588

### Bayes Theorem (Bayes teoremi)

<br>
<img src="p8.png" align = "left" style="width:500px;height:300px"/>

P(A|B) = B olayı gerçekleştiğinde A olayının gerçekleşme olasılığı

P(A) = A olayının gerçekleşme olasılığı

P(B|A) = A olayı gerçekleştiğinde B olayının gerçekleşme olasılığı

P(B) = B olayının gerçekleşme olasılığı

Şimdi bu formulün nasıl çıktığına bakalım

    P(A|B) = P(A and B) / P(B)

    P(B|A) = P(B and A) / P(A)

    P(A and B) = P(B and A)

    P(A|B).P(B)=P(B|A).P(A)

    P(A|B) = P(B|A) . P(A) / P(B)

Örnek:

Bir okul var. Bu okulda 50 tane çocuk olsun. Bu 50 çocuktan 1 tanesi x hastalığına yakalanabilir. x hastalığına sahip hastalıklı bir çocuğun testi %90 pozitif, sağlıklı bir çocuğun testi ise %10 pozitif sonuç verebilir. Test sonucu pozitif olan bir çocuğun gerçekten x hastalığına sahip olma olasılığı nedir?

    P(A) : Çocuğun hasta olması olasılığı = 1/50
    
    P(B) : Testin pozitif çıkması olasılığı = 1/50 0.90 + 49/50 0.10 = 0.116
    
    P(A|B) : Pozitif çıkan testin hastalık çıkma olasılığı ?

    P(B|A) : x hastalığına sahip çocuğun testinin pozitif çıkma olasılığı = 0.9
    
    P(A|B)=P(B|A)P(A)/P(B) => (0.9 1/50) / (0.116) = %15

<font color="Chocolate"><br>
https://github.com/raziyedogan/IntroductionToMachineLearning/blob/main/Naive-Bayes/NaiveBayes.ipynb ve https://github.com/raziyedogan/MachineLearning/blob/main/14-NaiveBayes/NaiveBayes.ipynb linklerinden machine learning kısmında Bayes teoremine dair detaylı bir anlatım bulunmaktadır. Buradan da bu konuyu inceleyebilirsiniz.